### PR TITLE
Add missing index on notified_project's project_id

### PR DESCRIPTION
**Pull request by danidoni**
The second slowest query in the slow log report that you can find in the attached [Trello card](https://trello.com/c/94X2HYZT/2215-not-indexed-queries) examines a lot of rows.

```
# Query 2: 0 QPS, 0x concurrency, ID 0x62E363BA6ED76266A41605B6B8D6CAC0 at byte 161791658
# This item is included in the report because it matches --limit.
# Scores: V/M = 0.00
# Time range: all events occurred at 2021-12-06 13:42:58
# Attribute    pct   total     min     max     avg     95%  stddev  median
# ============ === ======= ======= ======= ======= ======= ======= =======
# Count          0       1
# Exec time     20     16s     16s     16s     16s     16s       0     16s
# Lock time      0    97us    97us    97us    97us    97us       0    97us
# Rows sent      0       0       0       0       0       0       0       0
# Rows examine  49  54.15M  54.15M  54.15M  54.15M  54.15M       0  54.15M
# Rows affecte   0       0       0       0       0       0       0       0
# Bytes sent     0     373     373     373     373     373       0     373
# Query size     0     101     101     101     101     101       0     101
# String:
# Databases    frontend
# Hosts        ibs-proxy-1.ibsmgmt.suse.de
# Users        obs
# Query_time distribution
#   1us
#  10us
# 100us
#   1ms
#  10ms
# 100ms
#    1s
#  10s+  ################################################################
# Tables
#    SHOW TABLE STATUS FROM `frontend` LIKE 'notified_projects'\G
#    SHOW CREATE TABLE `frontend`.`notified_projects`\G
# EXPLAIN /*!50100 PARTITIONS*/
SELECT `notified_projects`.* FROM `notified_projects` WHERE `notified_projects`.`project_id` = 474649\G
```

Maybe we are not hitting any index for that query:
```
MariaDB [api_development]> explain SELECT `notified_projects`.* FROM `notified_projects` WHERE `notified_projects`.`project_id` = 24;
+------+-------------+-------------------+------+---------------+------+---------+------+------+-------------+
| id   | select_type | table             | type | possible_keys | key  | key_len | ref  | rows | Extra       |
+------+-------------+-------------------+------+---------------+------+---------+------+------+-------------+
|    1 | SIMPLE      | notified_projects | ALL  | NULL          | NULL | NULL    | NULL | 11   | Using where |
+------+-------------+-------------------+------+---------------+------+---------+------+------+-------------+
```
Great, this query is doing a full table scan (no possible keys listed and no key used after all). After the creation of the index, you can see that it hits the newly created index and the number of rows scanned are lower:
```
MariaDB [api_development]> explain SELECT `notified_projects`.* FROM `notified_projects` WHERE `notified_projects`.`project_id` = 24;
+------+-------------+-------------------+------+---------------------------------------+---------------------------------------+---------+-------+------+-------+
| id   | select_type | table             | type | possible_keys                         | key                                   | key_len | ref   | rows | Extra |
+------+-------------+-------------------+------+---------------------------------------+---------------------------------------+---------+-------+------+-------+
|    1 | SIMPLE      | notified_projects | ref  | index_notified_projects_on_project_id | index_notified_projects_on_project_id | 4       | const | 1    |       |
+------+-------------+-------------------+------+---------------------------------------+---------------------------------------+---------+-------+------+-------+
```

### CODE REVIEW

File: src/api/spec/db/data/backfill_notified_projects_spec.rb [hunk 1]
```diff
@@ -39,13 +39,16 @@
     end
 
     it 'backfills the notifications_projects table with all projects from existing notifications' do
-      expect(NotifiedProject.pluck(:project_id)).to eq([
+      expected = NotifiedProject.pluck(:project_id).sort
```
Comment (hennevogel) [inline src/api/spec/db/data/backfill_notified_projects_spec.rb:42]: Why is this change necessary?

**Comment by danidoni** [comment]
Referenced Trello card https://trello.com/c/94X2HYZT/2215-not-indexed-queries

File: src/api/spec/db/data/backfill_notified_projects_spec.rb (hunk 1 above)
Comment (danidoni) [reply src/api/spec/db/data/backfill_notified_projects_spec.rb:42]: Because the index somehow changed the order in which the results are returned and the spec didn't enforce any other order.

**Comment by hennevogel** [comment]
What about `.database_consistency.todo.yml`?

**Comment by danidoni** [comment]
> What about `.database_consistency.todo.yml`?



> What about `.database_consistency.todo.yml`?

I addressed a foreign key check for that column. Let's wait for the checks...

File: src/api/spec/db/data/backfill_notified_projects_spec.rb (hunk 1 above)
Comment (hennevogel) [reply src/api/spec/db/data/backfill_notified_projects_spec.rb:42]: Things like this need to be mentioned in the commit message...

Comment (danidoni) [reply src/api/spec/db/data/backfill_notified_projects_spec.rb:42]: On it

Comment (danidoni) [reply src/api/spec/db/data/backfill_notified_projects_spec.rb:42]: Done

File: src/api/db/migrate/20230414085150_add_notified_projects_project_id_index.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class AddNotifiedProjectsProjectIdIndex < ActiveRecord::Migration[7.0]
+  def change
+    add_index :notified_projects, :project_id
```
[the remark this answers is not in the document]
Comment (hennevogel) [reply src/api/db/migrate/20230414085150_add_notified_projects_project_id_index.rb:3]: Might be something we have to schedule for the maintenance window then? Did you check how long this takes?

```
[2] pry(main)> NotifiedProject.count
=> 4350704
```

[the remark this answers is not in the document]
Comment (danidoni) [reply src/api/db/migrate/20230414085150_add_notified_projects_project_id_index.rb:3]: [Strong migrations](https://github.com/ankane/strong_migrations#safe-by-default) says it's safe to add an index. So it should not lock the table and be safe to do it outside the maintenance window.

[the remark this answers is not in the document]
Comment (hennevogel) [reply src/api/db/migrate/20230414085150_add_notified_projects_project_id_index.rb:3]: Hm but we don't use that setting...

[the remark this answers is not in the document]
Comment (danidoni) [reply src/api/db/migrate/20230414085150_add_notified_projects_project_id_index.rb:3]: After reading [MariaBD documentation regarding adding indexes](https://mariadb.com/kb/en/innodb-online-ddl-operations-with-the-inplace-alter-algorithm/#alter-table-add-primary-key) we can ask for a specific locking mechanism to be used, but if that strategy cannot be applied, the migration will fail. If we don't ask for a specific locking mechanism the default one kicks in and then it tries to acquires the least restrictive one at runtime. So I think it's better to schedule this to be run during the maintenance window in case the MariaDB engine chooses to lock the table. WDYT?

[the remark this answers is not in the document]
Comment (hennevogel) [reply src/api/db/migrate/20230414085150_add_notified_projects_project_id_index.rb:3]: sounds about right

[the remark this answers is not in the document]
Comment (danidoni) [reply src/api/db/migrate/20230414085150_add_notified_projects_project_id_index.rb:3]: As for how long will it take, it depends on the algorithm the engine uses to perform the alter table... From ` incredibly slow` for the `ALGORITHM=COPY`, to almost instantaneous if `ALGORITHM=INSTANT` is chosen, but I haven't yet found how to tell ActiveRecord to let me choose an alter table algorithm...

**Comment by danidoni** [comment]
Rebased against master, nothing new.

File: src/api/spec/db/data/backfill_notified_projects_spec.rb [hunk 3]
```diff
@@ -39,13 +39,16 @@
...
+      real = [
         bs_request_with_submit_action.target_project_objects.distinct.map(&:id),
         user_review.bs_request.target_project_objects.distinct.map(&:id),
         comment_project.commentable.id,
         comment_package.commentable.project_id,
         comment_request.commentable.target_project_objects.distinct.map(&:id)
-      ].flatten!)
+      ].flatten!.sort
+
+      expect(expected).to eq(real)
```
Comment (eduardoj) [inline src/api/spec/db/data/backfill_notified_projects_spec.rb:51]:
```suggestion
      ].flatten!

      expect(expected).to contain_exactly(real)
```
Here we could use the [contain_exactly](https://rspec.info/features/3-12/rspec-expectations/built-in-matchers/contain-exactly/) matcher, so there is no need to use `.sort`.

**Review by eduardoj — APPROVED** [review]

**Comment by eduardoj** [comment]
Using the mainenance window. Removing the "DO NOT MERGE" label and merging.
